packages/doenetml-worker-rust/lib-doenetml-core/src/core/components/doenet/text_input/state_vars/immediate_value.rs

```diff
@@ -69,12 +51,10 @@ impl StateVarUpdater<String> for ImmediateValueStateVar {
     }
 
     fn invert(
-        &mut self,
+        data: &mut RequiredData,
```

Comment:
It would be great if we could find a way to make this immutable. Then we wouldn't have to think about whether the parent caller was doing something else with the data. Alternatively, can we consume `RequiredData`? Or does Core need to hold onto that?

Reply:
It needs to be mutable because the `StateVarView<T>` stores the requested value on itself (even though it is "read only" as far as core is concerned, it is a mutable variable as far as rust is concerned).

Reply:
But there could be another structure that we create, mutate, and then return. Isn't the only thing that's needed here is that the data is being saved just long enough to do `data.queued_update()`? Or will we access the `is_queued`/`requested_value` state from some other function? If the latter is the case, the doc string needs some bold text pointing out that the mutation has big consequences.

Reply:
A possible solution is if `RequiredData` had an `update_queue() -> RequiredDataUpdateQueue` function. Then we just make a mutable update queue in this function, use it to generate our `queued_updates()` and then throw it away.

Reply:
No, the way we communicate the value through the cached data is to directly set the requested value on the cached data. All we send to core is a list of dependencies to recurse to, not their desired values. (I knew you weren't going to like this once you discovered what was going on!)

Reply:
The reason for doing this is the same for using the typed, cached data: to avoid the need to pass data through core, where we'd need to wrap in enums and unwrap on the other end. We use the cached data (with its built in Rcs) to communicate in both directions, relying on the structure of `StateVarView<T>`, including its private fields and limited exposed API, to ensure data consistency.

Reply:
Wait a minute, the only thing that isn't behind a `RefCell` is the `update_has_been_queued` flag, the info of which does get thrown away after we generate our `queued_updates`, so maybe there is a way to avoid the `mut` as you suggested.

Reply:
I'm not immediately seeing a way to do this, but it seems like there should be a way to get rid of the mutability of `RequiredData`. Unless you see a quick way do to this, maybe make an issue rather than hold us this PR?

Reply:
Another PR sounds good. This one is getting big :-P